### DISCUSSION TITLE
プレビュー画面全体の印刷ができない

### DISCUSSION THREAD

**Discussion by tsubolabo** [increments/qiita-discussions#314] [要望・アイデア・不具合]
投稿が長文で複雑になると、画面上だけで推敲を加えるのは困難になります。ときどき、全体を見渡せる状態に印刷出力したいのですが、Qiita のプレビューではほんの一部分しか印刷されず困っています。

　そこで現状の対応として、印刷が必要なほどの長文の投稿の場合には、Qiitaのエディタを使わず、ブラウザの拡張機能 Markdown Viewer を見ながら汎用エディタで原稿を作り、最終段階で Qiita のエディタにコピペして仕上げるという方法を採っています。しかし、Markdown Viewer と Qiitaエディタ では文法が完全には同一でないため、コピペ後の手直しにかなり時間を食います（特に、数式のエスケープ方法の違いが厳しい）。

　そこで、お願いです。Qiita のプレビュー画面から、一つの記事の全文の印刷ができるようにしていただけないでしょうか？（用紙やインキの無駄がでないように、広告等は印刷されないように最適化されると助かります）

**Comment by tanaka-asami415** [comment]
@tsubolabo 
ありがとうございます🙏
恐れ入りますが、プレビューは印刷することを前提としていないため、
プレビュー画面の印刷状態を整える実装は現時点では考えておりません。

代替案として、一度、限定共有投稿したものを印刷していただければ、
推敲にご使用いただけるのではないかと思います。お試いただけますと幸いです🙇
限定共有投稿につきましては、こちらのヘルプをご参照ください。
https://help.qiita.com/ja/articles/qiita-private-article

**Reply by tsubolabo** [reply]
ご回答ありがとうございます。
限定共有投稿してからpdfに印刷してみました。
最初から最後までの印刷はできるようですが、各ページの境目で行の欠落があります（添付図は、原文と印刷欠落部のプレビュー画面のスクリーンコピーです）。
pdfファイルへの出力後の画像も添付図のプレビューと同じでした。

また、限定共有投稿ではない正規の投稿記事でも試してみましたが、ページの切れ目で同様に欠落が発生します。

<img width="300" alt="限定共有の印刷２" src="https://user-images.githubusercontent.com/42261605/195262033-d486a864-326f-4eb7-b4cb-6ec01f70204d.png">

<img width="882" alt="限定共有の印刷" src="https://user-images.githubusercontent.com/42261605/195262103-a0addef6-9209-4463-9dfc-9997141b23ed.png">

**Reply by tanaka-asami415** [reply]
@tsubolabo 
ご不便をおかけして申し訳ありません。
Qiitaは印刷することを想定していないため、境目が切れてしまう点はご容赦いただけますと幸いです🙇

**Reply by tsubolabo** [reply]
え～っ！！！

>代替案として、一度、限定共有投稿したものを印刷していただければ、
推敲にご使用いただけるのではないかと思います。お試いただけますと幸いです🙇

と言われて、喜び勇んで試してみたのに、その結果が

>Qiitaは印刷することを想定していないため、境目が切れてしまう点はご容赦いただけますと幸いです🙇

の一言では救われません。この印刷状態では推敲なんてできませんよね。

　しかし、Qiita上の目ぼしい記事が全部pdf化されて流出してしまえば、Qiitaにとってはとんでもない事態で、完璧な印刷ができないようにするのは正当な自衛手段とも思えます。しかし、投稿者本人しかアクセスできないプレビューページを印刷できるようにしても、Qiitaには何のダメージも無いように思います。

ということで、話は振り出しに戻りました。却下は覚悟のうえで、下記の提案を記録に留めさせていただきます。

「プレビュー画面の文書全体を欠損なく印刷する機能が欲しい」

**Reply by yoshinoyuya** [reply]
@tsubolabo 
「文書全体を欠損なく印刷する」ことだけでしたら、
PDF印刷の「その他の設定 > 用紙サイズ」をA4からA3の解決方法ではだめでしょうか

欠損が生まれる原因としては、レスポンシブデザイン(スマホ等の横幅が狭い場合)にUIが変わることが要因です。
そのため、A4より横幅が広いA3で欠損なく印刷が可能となっています。

例としてA3をあげましたが、横幅が広ければ、
欠損がなく印刷できることから、可能な他の用紙サイズもあると思います。

以下は設定変更による、プレビューの違いです。([印刷に使用したページ](https://qiita.com/Qiita/items/c686397e4a0f4f11683d))
| サイズ |  PDF印刷プレビュー(Edge)  |  実際のWebページ(Edge)  |
| ---- | ---- | ---- |
| A4 |  ![image](https://user-images.githubusercontent.com/85545443/195774387-2de80304-0585-4eb9-b53e-282ccc624eec.png)  |  ![image](https://user-images.githubusercontent.com/85545443/195774868-d8c63723-40e3-4586-8332-5937232b3ddf.png)  |
| A3 |  ![image](https://user-images.githubusercontent.com/85545443/195778045-e8166fc6-152d-4e7f-81f4-1810f25eb517.png)  |  ![image](https://user-images.githubusercontent.com/85545443/195778096-af926ab2-9fd2-4640-9b66-bb13508d5f21.png)  |

**Reply by tsubolabo** [reply]
[yoshinoyuya](https://github.com/yoshinoyuya) さん、貴重な情報をありがとうございます。
事情にお詳しいようですが、貴方はQiitaの方でしょうか？

ご教示いただいた方法を自分の投稿予定の限定共有投稿記事で試してみたところ、次のようになりました。推敲目的だけであれば十分に使えそうです。

+ AcrobatでpdfのA3縦に出力
　↓
行の欠落はなく全体が出力できました。ただし、一部、ページの切れ目で、一行の文字が上下に分割され、前ページと次ページに分かれてしまう部分がでてきます。しかし、なんとか判読は可能です。

+ A3サイズのpdfをインクジェットプリンタでA4に縮小印刷
　↓
文字はかなり小さくなりますが、なんとか読むことはできます。考えようによっては、紙数が減らせて経済的です。

A3縦以外にも、レターサイズ縦、Ｂ４縦も試してみましたが、欠落なく出力できるのはA3縦だけでした。

ところで、[tanaka-asami415](https://github.com/tanaka-asami415) さんの言われていた下記の文言は、私が問題にしているページの変わり目での行の欠落のことではなく、行の上下分割のことなのでしょうか？　見本を添付していたのに、うまく伝わっていなかったのでしょうか？

>Qiitaは印刷することを想定していないため、境目が切れてしまう点はご容赦いただけますと幸いです🙇

**Reply by yoshinoyuya** [reply]
@tsubolabo 
> 事情にお詳しいようですが、貴方はQiitaの方でしょうか？

ただのユーザーです。
A4で印刷する場合だと、自分が思いつく対策は「拡大率変更」か「CSS変更」の2つです。


## 拡大率変更
拡大率を100から下げて「サイズ69」の画像のように欠損がない拡大率を探します。
|  サイズ70  |  サイズ69  |
| ---- | ---- |
|  ![image](https://user-images.githubusercontent.com/85545443/196068002-24ec2c6e-fe94-4e7d-b923-9fce5a772d66.png)  |  ![image](https://user-images.githubusercontent.com/85545443/196067830-18a6de76-614c-4c99-8af4-669ae28acab9.png)  |





## CSS変更
CSSを変更して、下図の画面下部に出てくる「いいね・ストック」等の部分を消します。
|  CSS変更前  |  CSS変更後  |
| ---- | ---- |
|  ![image](https://user-images.githubusercontent.com/85545443/196068681-5ba27a92-e79b-4589-b515-edea20537df6.png)  |  ![image](https://user-images.githubusercontent.com/85545443/196068736-eae43ed9-7972-420c-a35c-bc2c72895c38.png)
  |





1. 該当ページにアクセスします
![image](https://user-images.githubusercontent.com/85545443/196066644-73eb4c55-d08e-4dd8-a9af-b91c3fa73288.png)
2. 右クリックをし、「開発者ツールで調査する」をクリックします。
![image](https://user-images.githubusercontent.com/85545443/196066517-0f235705-fa49-497f-a697-ebeec7141852.png)
3. 「コンソール」をクリックします。
![image](https://user-images.githubusercontent.com/85545443/196066848-1b9ddec3-edb1-4cd7-b20d-cf87d22e38e3.png)
4.以下の文字列をコピー・ペーストします。
```js
document.querySelector(`[id^='PersonalArticlePage-react-component'] > div:nth-child(2)`).style.display = "none"
```
![image](https://user-images.githubusercontent.com/85545443/196067194-c2439153-f071-49f2-8daf-e32aad4a86bb.png)
5.「Enter」を押し、実行します。
![image](https://user-images.githubusercontent.com/85545443/196067254-32a16b02-1981-41d3-b0ae-97514a0da533.png)
6. A4で印刷します。

以上がユーザーとして出来ることかと思います。

## 改善案
ここからは、Qiita側が印刷機能を追加すべきかという点を話していきたいと思います。

個人的な意見としては、要望するユーザーが多けば実装すべきだと思います。
ですが、Qiita記事を印刷して何かに利用するユーザーは少ないと思います。

そのため、改善するとすれば、編集画面のプレビュー画面だと思います。
自身の案だと、プレビュー画面を別ウィンドウ(youtubeコメント欄ようなイメージ)で開けるようにするとかだと思います。
このような感じで、Qiitaエディタの改善案を上げるのが良いかと思います。

[tsubolabo](https://github.com/tsubolabo)さんの場合ですと

>　投稿が長文で複雑になると、画面上だけで推敲を加えるのは困難になります。ときどき、全体を見渡せる状態に印刷出力したいのですが、Qiita のプレビューではほんの一部分しか印刷されず困っています。

`全体を見渡せる状態`はどのような表示なら該当するか

>　そこで現状の対応として、印刷が必要なほどの長文の投稿の場合には、Qiitaのエディタを使わず、ブラウザの拡張機能 Markdown Viewer を見ながら汎用エディタで原稿を作り、最終段階で Qiita のエディタにコピペして仕上げるという方法を採っています。

` Markdown Viewer`のどのような機能がQiitaエディタと違って書きやすいのかなど
を上げることで、Qiitaとしても動きやすいのかと思います。

**Reply by munierujp** [reply]
>自身の案だと、プレビュー画面を別ウィンドウ(youtubeコメント欄ようなイメージ)で開けるようにするとかだと思います。

FYI: はてなブログには、Qiitaのようにエディタに横並びでリアルタイムプレビューを表示する機能もありますが、それとは別に実際の記事の見た目に近い形でプレビュー表示する機能がありますね。

cf. https://help.hatenablog.com/entry/editor/preview

**Reply by tanaka-asami415** [reply]
@yoshinoyuya 
詳細にありがとうございます🙏

**Reply by tsubolabo** [reply]
[yoshinoyuya](https://github.com/yoshinoyuya) さん
詳細なご指導をありがとうございました。
お教えいただいた方法での実行結果を報告させていただきます。私にとっては、拡大率変更が最適な方法のように感じます。
+ **拡大率変更**

　用紙サイズをA4に設定し、拡大率を100%から徐々に減らして行きます。70%までは三点リーダがフッター部にあるので一部の文字が隠れています。69%にすると三点リーダがページの左端に移動し、隠される部分はなくなります。さらに%を減らして行くと文字サイズは当然小さくなっていきます。54%まで小さくなりページ幅に余裕が出てくると、ページの右端に目次まで現れるようになります。
　目次は不要なので、文字サイズができるだけ大きい69%に設定しておき、インクジェットプリンタに直接出力して印刷しました。A3のpdfからA4に縮小プリントしたときに比べると、印刷結果はより鮮明になりました。
　また、これは微妙なサイズ誤差による偶然の結果かもしれませんが、A3のときの行の上下分断が４ヶ所（全18ページに対して）だったのに比べ、今回は１ヶ所に減っています。本来なら、A3とA4の幅の比は1.41なのが、今回は1/0.69=1.45と少々異なるからでしょうか。

+ **CSS変更**

　この方法によれば、読みやすい文字サイズのままA4用紙に直接印刷できます（ページ数は当然増えます）。もちろん行の欠落はありませんが、行の上下分断も皆無で優秀です。しかし、数式に許される幅がかなり狭くなっているようで、本来は一行に収まるはずの数式が途中で強制的に折り返され、レイアウトがかなり乱れます。
　それに、開発者ツールに打ち込む document.querySelector(`...`).style.display = "none" の意味がまだ良く理解できておらず、私には少々敷居が高い気がします。

**次に御下問の２点についての回答です**
>全体を見渡せる状態はどのような表示なら該当するか

　長文の画面上の編集では、行きつ戻りつ頻繁にスクロールして目的の場所を見つけ、次々に処理していくのが結構大変です。印刷したものをページを繰りながら、赤ペンで訂正・加筆する方が遥かに楽です。「全体を見渡せる」というのは適当な表現ではなかったかもしれません。「印刷したものを全ページが見渡せるように床に広げる」わけではありません。ホッチキスで閉じた状態で使用します。

>Markdown Viewerのどのような機能がQiitaエディタと違って書きやすいのか

　Markdown Viewerは、書きやすいというよりも簡単に印刷できる点を買って利用していました。しかし、使ってみると書きやすさという点でも次のような良さを感じます。（Qiitaのエディタも、数式が無い短い記事であれば超便利だと思います。）
+ 長文の場合には、一定量を打ち込んでから一気に変換する方が落ち着ける
+ 数式の変換に癖がない。

しかし、Qiitaとは文法が若干異なる点に難はあります。Qiitaでも全文を印刷できることを知った今では、どちらが便利かを考え直してみる必要があります。

**Reply by tsubolabo** [reply]
もう、議論が尽くされた感もありますが、書き忘れたことがありますので追記します。

+ 拡大率変更印刷
　文章の一行が前後ページに分断して印刷される部分もあると書きましたが、図と表は、これとは少し様子が異なります。
　図は、そのページの残りのスペースが狭いときには、分断されないようにページ送りされ、図の全体が次ページの頭から始まります。
　表は分断されますが、分断された表の下半分の頭にはタイトル行が再挿入されます。

+ Marldown viewer の使い易さ
　使い慣れた汎用エディタで作業できるのも利点の一つです。Qiitaのエディタで作業しているとき、例えば、汎用エディタの「元に戻す」ボタンに相当するものがないので、勝手が違ってマウスカーソルの行き場に困ることがあります。（とは言っても、Qiitaのエディタの機能を増やして操作を複雑化して欲しいわけではありません。）
　他にMarkdown Preview Plusという拡張機能も試しましたが、数式がQiitaのMathJax方式ではなくKaTex方式だったので避けました。

**Comment by tanaka-asami415** [comment]
プレビュー画面については他にも多数ご意見をいただいており、改善を重ねていきます💪
